proposals/p2200.md

```diff
@@ -0,0 +1,1075 @@
+}
+// ...
+```
+
+Once all types have implemented the new interface, the blanket implementation
+can be removed:
+
+```carbon
+// Template constraint `HasF` no longer needed.
+
+// New interface
+interface NewF {
+  fn DoF[me: Self]();
+}
+
+// Blanket implementation no longer needed.
+
+fn G[template T:! NewF](x: T) {
+  x.DoF();
+}
+
+class C {
+  impl as NewF {
+    fn DoF[me: Self]();
+  }
+}
+var y: C = {};
+// `C` implements `NewF` directly.
+G(y);
+```
+
+The [name lookup rules](#name-lookup) ensure that unqualified names will only
+have one possible meaning, or the compiler will report an error. This error may
+be resolved by adding qualifications, which is done with the context of an
+ambiguity for a specific type. This avoids silent changes in the meaning of the
+code. Once the disambiguating qualifications have been added, the transition to
+checked generic becomes safe.
+
+#### To checked generic
+
+Once all needed qualifications are in place, the `template` keyword can be
+removed. After this, names will only be looked up in the interface constraints,
+not the type. If this does not cover the names used by the function, the
+compiler will report an error and this step can be rolled back and the previous
+step can be repeated to cover what was missed.
+
+Qualifications in the body of the function can be removed at this point, if
+desired, since the compiler will complain if this introduces an ambiguity from
+two different interfaces using that name. It would be reasonable to leave the
+qualifications in if the type parameter has multiple interface constraints, both
+as documentation for readers and to protect against future name collisions if
+the interfaces are changed.
+
+### Validity can depend on value
+
+A templated parameter may be used in ways where the validity of the result
+depends on the value of the parameter, not just its type. As an example, whether
+two array types are compatible depends on whether they have the same size. With
+a symbolic constant sizes, they will only be considered equal if the compiler
+can show that the two sizes are always equal symbolically. If the size is a
+template parameter, the checking will be delayed until the value of the template
+parameter is known.
+
+### Template dependent
+
+Expressions fall under three categories:
+
+-   Expressions that are valid and have meaning determined without knowing the
+    value of any template parameter are _not template dependent_.
+-   Expressions whose meaning and validity requires knowing the value of a
+    template parameter are _template dependent_.
+-   Expressions that have a meaning without knowing the value of any template
+    parameter, assuming it is valid, but whose validity requires knowing the
+    value of a template parameter are _template validity dependent_. These
+    expressions can trigger a monomorphization error, but are not considered
+    template dependent for purposes of a containing expression.
```

Comment:
How can I phrase this to be clearer? I think I was trying to get at what you were saying with "but are not considered template dependent for purposes of a containing expression". Maybe there is an example that would clarify?

Reply:
I added a paragraph https://github.com/carbon-language/carbon-lang/pull/2200/commits/1a670dd5e5bf64688956a681023910479cc23c39 , does that address your concerns?

Reply:
I think it's now clear that "template dependent" refers only to properties of the "immediate" expression and not to its subexpressions. But I still don't know whether template validity dependence takes subexpression validity into account or not. Given:
```
fn G(n: i32) -> i32;
fn F[template T:! Type](x: T) -> i32 { return G(x.y); }
```
... is the expression `G(x.y)` template validity dependent, because its validity depends on whether `T` has a member `y`, or is it not template dependent because we know that it is an rvalue of type `i32` that calls `G` regardless of the properties of `T`?

I don't think this needs to be resolved immediately -- we're not using this notion for much, yet -- but we'll need an answer once we start using this terminology to drive other language rules.